www/blog/2023-01-14-typescript-performance-lessons.md

```diff
@@ -0,0 +1,195 @@
+
+For tRPC, we do our best to ensure this by [generating](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/scripts/generateBigBoi.ts) and [testing](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/packages/tests/server/react/bigBoi.test.tsx) a router with 3,500 procedures and 1,000 routers. But this only tests how far we can push the TS compiler before it breaks and not how long type-checking takes. We test all three pieces of the library (server, vanilla client, and the React client) because they all have different code paths. In the past, we have seen regressions that are isolated to one section of the library and rely on our tests to show us when those unexpected behaviors occur. (We still do [want to do more](https://github.com/trpc/trpc/issues/2892) to measure compilation times)
+
+tRPC is not a runtime-heavy library so our performance metrics are centered around type-checking. Therefore, we stay mindful of:
+
+- Being slow to type-check using `tsc`
+- Having a large initial load time
+- If the TypeScript language server takes a long time to respond to changes
+
+The last point is one that the tRPC must pay attention to the most. You **never** want to your developers to have to wait for the language server to update after a change. This is where tRPC **must** maintain performance so that you can enjoy great DX.
+
+## How I found performance opportunities in tRPC
+
+There is always a tradeoff between TypeScript accuracy and compiler performance. Both are important concerns for other developers so we must be extremely conscious of how we write types. Will it be possible for an application to run into severe errors because a certain type is "too loose"? Is the performance gain worth it?
+
+Is there even going to be a meaningful performance gain at all? Great question.
+
+Let's have a look at _how_ to find moments for performance improvements in TypeScript code. We'll visit the process I went through to create [PR #2716](https://github.com/trpc/trpc/pull/2716), resulting in a 59% decrease in TS compilation time.
+
+---
+
+TypeScript has a built-in [tracing tool](https://github.com/microsoft/TypeScript/wiki/Performance-Tracing) that can help you find the bottleneck in your types. It's not perfect, but it's the best tool available.
+
+It's ideal to test your library on a real-world app to simulate what your library is doing for real developers. For tRPC, I created a basic [T3 app](https://create.t3.gg/) resembling what many of our users work with.
+
+Here's the steps I followed to trace tRPC:
+
+1. [Locally link the library](https://docs.npmjs.com/cli/commands/npm-link) to the example app. This is so you can change your library code and immediately test changes locally.
```

Comment:
Not sure why but this looks messed up in the preview
<img width="761" alt="CleanShot 2023-01-14 at 23 34 59@2x" src="https://user-images.githubusercontent.com/459267/212500138-e86029b1-b31c-4f89-9b85-ed6f8a6e7199.png">